src/commands/signzone.rs

```diff
@@ -0,0 +1,4157 @@
+        // add support for a -q (--quiet) option or similar but that was never
+        // done.
+        match self.iterations {
+            500.. => Self::write_extreme_iterations_warning(&env),
+            100.. if self.invoked_as_ldns => Self::write_large_iterations_warning(&env),
+            1.. if !self.invoked_as_ldns => Self::write_non_zero_iterations_warning(&env),
+            _ => { /* Good, nothing to warn about */ }
+        }
+
+        // Read the zone file.
+        let mut records = self.load_zone(&env.in_cwd(&self.zonefile_path))?;
+
+        // Extract the SOA RR from the loaded zone.
+        let Some(soa_rr) = records.find_soa() else {
+            return Err(format!(
+                "Zone file '{}' does not contain a SOA record",
+                self.zonefile_path.display()
+            )
+            .into());
+        };
+        let ZoneRecordData::Soa(_) = soa_rr.first().data() else {
+            return Err(format!(
+                "Zone file '{}' contains an invalid SOA record",
+                self.zonefile_path.display()
+            )
+            .into());
+        };
+
+        let dnskey_rrset = records.find_apex_rtype(soa_rr.owner(), Rtype::DNSKEY);
+
+        let cds_rrset = records.find_apex_rtype(soa_rr.owner(), Rtype::CDS);
+        let cdnskey_rrset = records.find_apex_rtype(soa_rr.owner(), Rtype::CDNSKEY);
+
+        // Extract and validate the DNSKEY RRs from the loaded zone.
+        let mut found_public_keys = vec![];
+        for rr in records.iter() {
+            if let ZoneRecordData::Dnskey(dnskey) = rr.data() {
+                // Create a public key object from the found DNSKEY RR.
+                let public_key = Record::new(rr.owner(), Class::IN, Ttl::ZERO, dnskey);
+
+                found_public_keys.push(public_key);
+            }
+        }
+
+        // Load the specified private keys, match them against the found
+        // public keys, failing that load a DNSKEY RR from the corresponding
+        // public key file and validate that its owner matches that of the
+        // zone apex. Unlike ldns-signzone we don't use a generated public key
+        // if these attempts fail.
+        let mut signing_keys: Vec<SigningKey<Bytes, KeyPair>> = vec![];
+
+        'next_key_path: for key_path in &self.key_paths {
+            let key_path = env.in_cwd(key_path).into_owned();
+            // Load the private key.
+            let private_key_path = Self::mk_private_key_path(&key_path);
+            let private_key = Self::load_private_key(&env.in_cwd(&private_key_path))?;
+
+            // Note: Our behaviour differs to that of the original
+            // ldns-signzone because we are unable at the time of writing to
+            // generate a public key from a private key. As such we cannot
+            // compare the key tag of any found DNSKEY RRs to that of the
+            // public key generated from the private key. Instead we attempt
```

Comment:
This comment appears to be wrong. It says for a single "found public key" and for "each private key" yet the following loop does the opposite trying to match each found public key to the one current private key.

Reply:
@Philip-NLnetLabs: Do you concur that this comment is incorrect?